### DISCUSSION TITLE
Angular xState example

### DISCUSSION THREAD

**Discussion by peterhe2000** [statelyai/xstate#1196] [General]
Thanks for a awesome package. Thanks for the great work. 
I am also a huge fans of FSM.
However I am struggling to find a good example of Angular working with xState. 
Is there a recommendation? 

List i found so far. Not many angular xstate example for some reason.
Seems each of those example got it's own implementation. 
Hope there is a standard which xstate team may suggest with.
https://stackblitz.com/edit/xstate-angular-table
https://stackblitz.com/edit/angular-xstate-http?file=src%2Fapp%2FloadMachine.ts
https://medium.com/angular-athens/working-with-state-machines-in-angular-2817441e26bf

**Comment by timdeschryver** [comment]
I think a simple machine can be implemented with the [RxJS recipe](https://xstate.js.org/docs/recipes/rxjs.html).

I also experimented with XState in Angular and came up with a [POC implementation](https://gist.github.com/timdeschryver/dcdeefdba43bef7134c8249e8257dae7) that was inspired by the `useMachine` hook from `xstate/react`. This implementation also takes a stream of events, instead of invoking the statemachine with the `send` method.

**Comment by peterhe2000** [comment]
Thanks @timdeschryver  Most of angular example seems come from you.
For ur POC implementation, how do i send machine next event from XStatePocComponent? 
Also where can i find ur machine which i can play in https://xstate.js.org/viz/?

**Comment by timdeschryver** [comment]
@peterhe2000 sending events is done via ["input event observables"](https://gist.github.com/timdeschryver/dcdeefdba43bef7134c8249e8257dae7#file-table-component-ts-L58).
Seems like I didn't add the machine to the gist, I don't have that code anymore - sorry.

**Comment by stefumies** [comment]
It's incredible, it's 2023 now and still the examples of using XState in Angular are lacking. They are either overly complicated, or lack the actual example of Angular implementation. It's quite frustrating. React has many great simple examples, but Angular seems to bring out the rocket scientist in people.

**Reply by davidkpiano** [reply]
Sorry, I would create examples but I do not have Angular experience. Is there an example that you've seen that is close to being acceptable (not overly complicated)? We can try to work from that.

**Reply by stefumies** [reply]
Unfortunately not that I can find. As I say there are  plenty of React examples, such as the basic Traffic light system. Something such as this as an Angular example would be desirable. 

Such a case  could provide for a simple example of integration with RXJS and Xstate with Angulars service led architecture, and using it's directives in the template interacting with the state for DOM activity and CSS. 

This was not a moan at you or anyone in particular but a general frustration at this absence, based on my increasing love of Xstate, it is a fantastic library, and I appreciate the work and effort gone into it. Thank you kindly for the prompt reply also!

**Reply by FloNeu** [reply]
Maybe we could make a bit of a combine effort... i am planing to use xstate for remodelling rebuilding some existing angular projects - no ready examples yet while i am still learning, but totally whilling to share and work on some stuff. For integrating the machine wit rxjs/signals (aka the new shit ^^) i found pretty straignt forward. 

The methodes available from an interpreted machine are from the top of my head (so correct me pls) **onChange**, **onDone**, **onEvent**, **onSend**, **onEvent**, **onStop**, **onTransitions** and **subscribe**.

The documentation under https://xstate.js.org/docs/recipes/rxjs.html is of course very missleading as it states: _The [interpreted machine](https://xstate.js.org/docs/guides/interpretation.html) (i.e., service) is subscribable._ which makes it sound like this would already be an observable according to rxjs implementation. 

That's of course not true - all the mentioned methodes are merely methodes taking a callback-function which gets the according machine outputs as parameters.

To integrate that with rxjs i create a BehaviourSubject for each methode and in the callback i pass to each forward the according machine outputs, like: 

```
@Injectable({
  providedIn: 'root'
})
export class SpinnerMachineV2Service {
  // I am not a 100% setteled on this three lines below... so take this with a grain of salt...
  // as initial state is probably not really what i want to initialze with... maybe i will change that to
  // just initialize with undefined... 
  public machine = interpret(SpinnerMachineV2, { devTools: true }); // Interprete the machine
  public state$ = new BehaviorSubject(this.machine.initialState); // Observable
  public stateS = signal(this.machine.initialState); // Signal

  constructor() {
    this.machine.subscribe( state => { // This is the subscribe-methode of the machine - not an rxjs observable
      this.state$.next( state ); // push state on the BehaviourSubject
      this.stateS.set( state ); // push state on the signal
    });
    this.machine.start();
  }
}
```

Just do a similar thing for the other methodes (i will probably to that in a BaseService i will extend for my machine services in the future)... Now i can this service in all places i require it via angular Dependecy injection and bind the properties i am interrested in my template like
```
<ng-container *ngIf="($spinner.state$ | async) as state">
  <div *ngIf="state.matches('SHOWN')">XSTATE in state SHOWN</div>
  <div *ngIf="state.matches('HIDDEN')">XSTATE in state HIDDEN</div>
<ng-container>
```
or just as you would with any other rxjs-observable/signal in code etc.
